tests/bit_cast/bit_cast_test.h

```diff
@@ -0,0 +1,109 @@
+//  limitations under the License.
+//
+//  Provides test for the sycl::bit_cast.
+//
+*******************************************************************************/
+#ifndef SYCL_CTS_BIT_CAST_TEST_H
+#define SYCL_CTS_BIT_CAST_TEST_H
+
+#include "bit_cast_helper_functions.h"
+#include <cstring>
+
+namespace bit_cast::tests {
+using namespace bit_cast::tests::helper_functions;
+
+constexpr int expected_val = 42;
+
+template <typename ToType, typename FromType>
+class kernel_name;
+
+template <typename ToType, typename FromType>
+class bit_cast_test {
+  static constexpr bool is_invalid_test_case =
+      std::is_array_v<ToType> || sizeof(ToType) != sizeof(FromType) ||
+      std::is_same_v<ToType, bool>;
+
+ public:
+  void operator()(const std::string& to_type_name,
+                  const std::string& from_type_name) {
+    auto queue = sycl_cts::util::get_cts_object::queue();
+    if constexpr (is_invalid_test_case)
+      return;
+    else {
+      std::array test_result{false, false};
+      {
+        sycl::buffer<bool, 1> buf_result(test_result.data(),
+                                         sycl::range<1>(test_result.size()));
+        queue
+            .submit([&](sycl::handler& cgh) {
+              auto result_acc =
+                  buf_result.template get_access<sycl::access_mode::write>(cgh);
+              cgh.single_task<kernel_name<ToType, FromType>>([=] {
+                FromType expected;
+                value_operations::assign(expected, expected_val);
+                FromType from;
+                value_operations::assign(from, expected_val);
+                auto to = sycl::bit_cast<ToType>(from);
+                result_acc[0] = std::memcmp(&to, &from, sizeof(from)) == 0;
+                if constexpr (!std::is_array_v<FromType>) {
+                  from = sycl::bit_cast<FromType>(to);
+                  result_acc[1] = value_operations::are_equal(from, expected);
+                }
+              });
+            })
+            .wait_and_throw();
+      }
+      {
+        INFO("Memory contents are not equal. "
+             << "ToType : " << to_type_name << " FromType: " << from_type_name);
+        CHECK(test_result[0]);
+      }
+      if constexpr (!std::is_array_v<FromType>) {
+        INFO("Round trip conversion failed. "
+             << "ToType : " << to_type_name << " FromType: " << from_type_name);
+        CHECK(test_result[1]);
+      }
+    }
+  }
+};
+
+template <typename PrimaryTypeTo, typename PrimaryTypeFrom>
+class run_bit_cast_test {
+ public:
+  void operator()(const std::string& primary_to_type_name,
+                  const std::string& primary_from_type_name) {
+    const auto to_types =
+        get_derived_type_pack<PrimaryTypeTo>(primary_to_type_name);
+    const auto from_types =
+        get_derived_type_pack<PrimaryTypeFrom>(primary_from_type_name);
+    const auto to_types_ptrs =
+        get_derived_type_pack<PrimaryTypeTo*>(primary_to_type_name + "*");
```

Comment:
From the spec (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:marray.type):
> The element type parameter, DataT, must be a numeric type as it is defined by C++ standard.

I might have missed it, but I don't think C++ defines "numeric" types. ComputeCpp fails CI because of this line, because it requires the data type to be `std::is_arithmetic` (or boolean). So a pointer will definitely not work, but again not sure if this is the intention of the spec.

Reply:
_NumericType_ is a C++ named requirement

https://en.cppreference.com/w/cpp/named_req/NumericType

Note that this is very broad, encompassing more than just arithmetic types.

Reply:
Thank you, I wasn't aware of that. In that case we missed it in ComputeCpp, for now we probably want to disable the test. Might also be good to clarify to say `NumericType` in the spec, a simple search for "c++ numeric type" surprisingly didn't find that requirement.

Reply:
I added this as a comment to an existing issue about `marray` and _NumericType_: https://gitlab.khronos.org/sycl/Specification/-/issues/559